### PR TITLE
[libcxx] reorganises the hardening documentation

### PR DESCRIPTION
**Pull request by cjdb**
The reorganisation assists with identifying information that's relevant to the reader by using sections, note/warning blocks, and highlighted lists.

Some rewording was necessary to fit the new structure and some to improve flow. Changes to the intention of the documentation have not been made.

### CODE REVIEW

**Comment by firewave** [comment]
In the past I mixed up defines used to _build libc++ itself_ and ones I can specify when _building against libc++_. It seems this is no longer possible but reading it I still didn't feel confident this is the documentation that relates to me until I read the very last paragraph. Maybe that could be a bit more emphasized with an introduction paragraph for the vendor/user section instead - maybe even using the phrases I highlighted in my first sentence.

**Comment by firewave** [comment]
I guess my confusion also stems a bit from the page which links to the hardening settings: https://libcxx.llvm.org/UsingLibcxx.html#libc-configuration-macros. That also contains some ambiguity with macros that extend to both cases. Maybe that also needs to better differentiate between those.

As that refers to "using libc++" I wonder if it should even mention anything related to the vendor/`building libc++ itself` case at all. Maybe that should be moved to a separate (and possibly already existing) page.

In the past it was why more confusing with assertions, annotations, debug mode, safe mode, etc. The new hardening stuff is definitely more straight forward. 👍

**Comment by cjdb** [comment]
@firewave would a table of contents help in that regard? Readers can then see that there's a section for them regardless of whether they're a user or a vendor.

Perhaps we need to reorganise the documentation so that it is bifurcated for users, maintainers, and vendors, but I'd prefer to do that in a separate, more general CL.

File: libcxx/docs/Hardening.rst
Comment (hawkinsw) [inline libcxx/docs/Hardening.rst:24]:
```suggestion
  some additional checks for undefined behavior that incur relatively little
```

File: libcxx/docs/Hardening.rst
Comment (hawkinsw) [inline libcxx/docs/Hardening.rst:25]:
```suggestion
  overhead but aren't security-critical. While the performance penalty in Extensive mode is
```

File: libcxx/docs/Hardening.rst
Comment (hawkinsw) [inline libcxx/docs/Hardening.rst:22]:
```suggestion
  used in production. We recommend most projects to adopt the Fast mode.
```

File: libcxx/docs/Hardening.rst
Comment (hawkinsw) [inline libcxx/docs/Hardening.rst:23]:
```suggestion
- **Extensive mode**, which contains all the checks from the Fast mode and
```

File: libcxx/docs/Hardening.rst
Comment (hawkinsw) [inline libcxx/docs/Hardening.rst:26]:
```suggestion
  somewhat more significant compared to the Fast mode, the Extensive mode is
```

File: libcxx/docs/Hardening.rst
Comment (hawkinsw) [inline libcxx/docs/Hardening.rst:27]:
```suggestion
  still intended for use in production builds.
```

File: libcxx/docs/Hardening.rst
Comment (hawkinsw) [inline libcxx/docs/Hardening.rst:39]:
```suggestion
As a libc++ user, consult with your vendor to determine the level of hardening
```

File: libcxx/docs/Hardening.rst
Comment (hawkinsw) [inline libcxx/docs/Hardening.rst:77]:
```suggestion
The ``LIBCXX_HARDENING_MODE`` option controls both the hardening mode that the precompiled library is
```

**Review by hawkinsw — COMMENTED** [review]
Thank you, @cjdb !! This documentation is very well written, succinct and helpful! Having read it I know much more about the different hardening levels. I hope that these suggestions are somewhat helpful!

File: libcxx/docs/Hardening.rst
Comment (cjdb) [reply libcxx/docs/Hardening.rst:25]: I think this one is already apparent from context.

File: libcxx/docs/Hardening.rst
Comment (cjdb) [reply libcxx/docs/Hardening.rst:77]: I think this one is already apparent from context.

**Comment by firewave** [comment]
> @firewave would a table of contents help in that regard? Readers can then see that there's a section for them regardless of whether they're a user or a vendor.

Not sure. I guess I have to look at the documentation from the building side.

The hardening page should probably just contains all the information on that (since that is generic) and be linked from the "using" and "building" (if there is one) pages with all information on vendor/user stuff possibly contained in those.

> Perhaps we need to reorganise the documentation so that it is bifurcated for users, maintainers, and vendors, but I'd prefer to do that in a separate, more general CL.

Just reducing any ambiguity is always helpful. I did not expect this to happen in this PR. Just saw the changes and remembered that I had issues with the documentation so I thought I drop my two cents.

**Comment by var-const** [comment]
@cjdb Thank you for working on this! Sorry I'm slow with the review (a combination of holidays and being OOO sick). LGTM modulo one comment (feel free to ping me on Discord if you'd like to discuss it further).

**Review by var-const — APPROVED** [review]

File: libcxx/docs/Hardening.rst
Comment (hawkinsw) [reply libcxx/docs/Hardening.rst:77]: Makes sense!!

File: libcxx/docs/Hardening.rst
Comment (hawkinsw) [reply libcxx/docs/Hardening.rst]: Ah, I see what you are saying, @var-const ! If I hear you correctly, you are saying that the rephrasing makes it seem like a recommendation to use this mode whereas the original phrasing has more of a "using this mode will not fail."

File: libcxx/docs/Hardening.rst
Comment (philnik777) [reply libcxx/docs/Hardening.rst:22]: I'm not a native speaker, but I think you use it as a proper noun in this context. The name of the mode is "Fast". It's not about the mode being fast.

Comment (hawkinsw) [reply libcxx/docs/Hardening.rst:22]: That was exactly what I was trying to say. Thanks for chiming in, @philnik777 !!

Comment (cjdb) [reply libcxx/docs/Hardening.rst:22]: It's a compound noun, not a proper noun.

Perhaps a different context will help: if we were recommending trees instead of hardening modes, it would be "We recommend most projects adopt the gum tree", not "We recommend most projects adopt the Gum tree".

Comment (philnik777) [reply libcxx/docs/Hardening.rst:22]: My understanding is that that still depends on the context. If we have an individual tree named "Gum", we would recommend the Gum tree, not the gum tree. This can probably be debated, but in the end I don't actually care that much. I haven't studied English and don't plan to. I can't even speak my native language properly.

Comment (hawkinsw) [reply libcxx/docs/Hardening.rst:22]: > My understanding is that that still depends on the context. If we have an individual tree named "Gum", we would recommend the Gum tree, not the gum tree. This can probably be debated, but in the end I don't actually care that much. I haven't studied English and don't plan to. I can't even speak my native language properly.

I'm with @philnik777 100%. We've already "wasted" way too much time here! 

> It's a compound noun, not a proper noun.
> 
> Perhaps a different context will help: if we were recommending trees instead of hardening modes, it would be "We recommend most projects adopt the gum tree", not "We recommend most projects adopt the Gum tree".

I agree with this particular example. However, I think that it's slightly inapplicable to the example here. But, again, we've already done way too much bikeshedding here! 

Thanks for the work, @cjdb !

Comment (var-const) [reply libcxx/docs/Hardening.rst:22]: (also not a native speaker) FWIW, I think this is purely a matter of style whether to capitalize the names of the modes. I personally prefer to mark them as code using Markdown when supported, while I find capitalization distracting.

Comment (hawkinsw) [reply libcxx/docs/Hardening.rst:22]: Thanks for weighing in @var-const -- that was the reason I suggested capitalization. I just wanted them to stand out a little!

**Pull request merged**
